backend/hasher/bolt.go

```diff
@@ -0,0 +1,371 @@
+// +build !plan9,!js
+
+package hasher
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/fs/config"
+	"github.com/rclone/rclone/fs/hash"
+	"github.com/rclone/rclone/fs/operations"
+	"github.com/rclone/rclone/lib/encoder"
+	"go.etcd.io/bbolt"
```

Comment:
>  I was worried about the project being maintained.

Bolt is heavily used by etcd which is in turn used by k8s which is a mature industry standard. I don't think they can deprecate it easily.

> can't be open from multiple processes at once which can be quite annoying

Bolt can be shared if everyone just want to read.
If someone wants to write, others should wait.
So each instance should handle locking and switch access mode.

Shared SQL databases (postgres, oracle) solve locking problem by using client-server pattern.
I solved it by using a serving goroutine. It handles locking and switches access mode to guarantee other instances a time slot. Other parts of rclone access database via API provided by the goroutine (I made an exception for the dump operation). I reduced core API to one function `do` and one `Op` interface. Everything database-related extends these two.

I considered a few alternatives
- use external KV server (e.g. redis is widespread and has very small go client)
- write a copy of VFS layer (fine grained per-file access but too much work)
- factor parts of VFS code (but it's a moving target)
- write my own KV store with locking (too much work)
- write my own KV store without locking and extend RCD REST protocol (or implement minimal redis server) so one rclone can serve KV to others (but I don't like revolutions and paradigm shifts)
- use bolt (it has transactions and locking, KV access needs some kind of locking anyway, transactions good though optional, so I chose bolt)

I don't see a single best choice. I chose something satisfactory.
In the end, I don't worry much. It's a cache. We can break and completely replace its implementation. In the worst case users will not loose data, just experience a delay.

A year ago I thought about metadata store for rclone. It's a bigger problem. It's not a cache, it can't loose data. It can't be recreated on every client machine. It needs either (a) a central store e.g. redis for all rclone instances or (b) use backends as golden sources of their metadata. (a) is simpler to impelement. I tried to tackle (b) i.e. sketch a KV storage engine that can use backends without modtime or locking for a consistent storage. Please take a look at [my design sketch](https://github.com/orgs/rclone/teams/rclone/discussions/1). It's a lot of work.

Reply:
P.S. A year ago I thought about _metadata store_ for rclone. It's a bigger problem. It's not cache - it can't lose data. It can't be recreated on every client machine.
It needs either (a) a central store e.g. redis for all rclone instances or (b) use backends as golden sources of their metadata.
(a) is simpler to impelement. I tried to tackle (b), sketch a KV storage engine that can use backends _without modtime or locking for consistent_ storage. Please take a look at [my design sketch](https://github.com/orgs/rclone/teams/rclone/discussions/1).